src/qutip_qip/vqa.py

```diff
@@ -0,0 +1,841 @@
+            If a list of floats of dimensions (n,) for n free parameters
+            in the circuit is given, then these are taken to be the initial
+            conditions for the optimizer. Otherwise if a string is given:
+            1. "random" will randomize initial free parameters between 0 and 1,
+            2. "ones" will set each initial free parameter to a value of 1.
+            Defaults to "random"
+        method: str or callable, optional
+            Method to give to ``scipy.optimize.minimize``
+        use_jac: bool, optional
+            Whether to compute the jacobian or not. If computed, it will be
+            passed to the optimizer chosen by ``method``, regardless of if
+            the method is gradient-based or not.
+            Note that derivatives of unitaries generated by
+            ``ParameterizedHamiltonian`` are calculated with the
+            Frechet derivative of the exponential function,
+            using ``scipy.linalg.expm_frechet``.
+        do_nothing: bool, optional
+            For debugging and benchmarking purposes, perform no operation
+            during optimization, and simply retain the initial free parameters.
+        layer_by_layer: bool, optional
+            Grow the circuit from a single layer, to ``VQA.n_layers``, at each
+            step holding the parameters found for previous layers fixed.
+        bounds: sequence or `scipy.optimize.Bounds`, optional
+            Bounds to be passed to the optimizer. Either
+
+            1. Instance of `scipy.optimize.Bounds`
+            2. Sequence of ``(min, max)`` tuples corresponding to each
+               free parameter.
+        constraints: list of `Constraint`
+            See `scipy.optimize.minimize` documentation.
+        """
+
+        n_free_params = self.get_free_parameters()
+        # Set initial circuit parameters
+        if isinstance(initial, str):
+            if initial == "random":
+                angles = [random.random() for i in range(n_free_params)]
+            elif initial == "ones":
+                angles = [1 for i in range(n_free_params)]
+            else:
+                raise ValueError("Invalid initial condition string")
+        elif isinstance(initial, list) or isinstance(initial, np.ndarray):
+            if len(initial) != n_free_params:
+                raise ValueError(
+                    f"Expected {n_free_params} initial parameters"
+                    f"but got {len(initial)}."
+                )
+            angles = initial
+        else:
+            raise ValueError(
+                "Initial conditions were neither a list of values"
+                " nor a string specifying initialization."
+            )
+
+        # Return early if do_nothing flag was passed
+        if do_nothing:
+            return self.evaluate_parameters(angles)
+
+        # Function for computing jacobian
+        jac = self.compute_jac if use_jac else None
```

Comment:
This is only used if `layer_by_layer==False`

Reply:
Good catch, thanks.